### PR TITLE
NIFI-12019 bugfix for SynchronousFileWatcher time check interval

### PR DESCRIPTION
**Pull request by mosermw**
# Summary

[NIFI-12019](https://issues.apache.org/jira/browse/NIFI-12019)  This PR corrects SynchronousFileWatcher so that it resets its last checked timestamp whenever it uses the configured UpdateMonitor to check on the file being watched.

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 17

### Licensing

- [n/a] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [n/a] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [n/a] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/monitor/SynchronousFileWatcher.java [hunk 1]
```diff
@@ -77,21 +77,16 @@ public boolean checkAndReset() throws IOException {
...
-                final StateWrapper wrapper = lastState.get();
+                final Object oldState = lastState.get().getState();
                 final Object newState = monitor.getCurrentState(path);
-                if (newState == null && wrapper.getState() == null) {
+                if (newState == null && oldState == null) {
                     return false;
                 }
-                if (newState == null || wrapper.getState() == null) {
-                    lastState.set(new StateWrapper(newState));
+                lastState.set(new StateWrapper(newState));
```
Comment (mattyb149) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/monitor/SynchronousFileWatcher.java:85]: If newState is `null` does this cause issues creating a wrapper around it?

Comment (mosermw) [reply nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/monitor/SynchronousFileWatcher.java:85]: I don't think so?  StateWrapper is a simple pojo with getters/setters and no real logic, so I don't think it cares.  This method handles checking for nulls on newState and oldState, so I was happy that NPE wasn't possible.

Comment (mattyb149) [reply nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/monitor/SynchronousFileWatcher.java:85]: Why the null checks then? Just to be defensive?

Comment (mosermw) [reply nifi-commons/nifi-utils/src/main/java/org/apache/nifi/util/file/monitor/SynchronousFileWatcher.java:85]: We don't know what UpdateMonitor implementation is currently in use, and that interface doesn't guarantee that `monitor.getCurrentState(path)` returns NotNull.  So we have to check for nulls before using `newState.equals()` on line 89.

**Comment by mattyb149** [comment]
+1 LGTM, thanks for the fix! Merging to support/nifi-1.x and main

**Comment by mosermw** [comment]
I appreciate you looking at this, @mattyb149.

**Comment by joewitt** [comment]
@mattyb149 @mosermw I had never even heard of this class.  But today I noticed this test failure https://github.com/joewitt/nifi/actions/runs/6252186486/job/16974874215

Can you please look into this?  I had never noticed this class failing before.

**Comment by mosermw** [comment]
@joewitt I'm looking at this now.  SynchronousFileWatcher starts a background thread to check if a file has changed, so the unit test timing tolerances are likely too tight.  Not sure if there's a good pattern for testing things like this that remove timing issues from the test.

**Pull request closed without merging**